### PR TITLE
[wallet] fix zapwallettxes interaction with persistent mempool

### LINKED ISSUE #9710: How should -zapwallettxes interact with persistent mempool?
the `zapwallettxes.py` qa test is currently broken due to a syntax error:

```python
     assert_raises(JSONRPCException, self.nodes[0].gettransaction, [txid3])
     #there must be a expection because the unconfirmed wallettx0 must be gone by now
```

This doesn't actually test that the transaction is not in the wallet, because `gettransaction()` is being called incorrectly (with an array instead of a string).

Changing the test to `assert_raises(JSONRPCException, self.nodes[0].gettransaction, txid3)` causes the rpc to succeed and the test script to fail because `txid3` is still in the wallet.

This is because the mempool is now persistent and is saved to disk when bitcoind shuts down. zapwallettxes removes the unconfirmed transaction from the wallet but bitcoind will rescan the mempool.dat file on startup and add the unconfirmed transaction back into the wallet.

I've 'fixed' zapwallettxes.py in https://github.com/jnewbery/bitcoin/tree/zapwallettxes , but the question remains: how *should* -zapwallettxes interact with persistent mempool? The use case for zapwallettxes is explained here: https://github.com/bitcoin/bitcoin/pull/3659#issuecomment-34878999 . My view is that even with zapwallettxes, bitcoind should rescan the mempool, but there should be better documentation or a warning that unconfirmed transaction in your mempool will be re-added to the wallet even if zapwallettxes is used. If you really want to remove all unconfirmed transactions from your wallet, you can delete your mempool.dat file and then restart with zapwallettxes.

### PR DESCRIPTION
**Pull request by jnewbery**
zapwallettxes previously did not interact well with persistent mempool.
zapwallettxes would cause wallet transactions to be zapped, but they
would then be reloaded from the mempool on startup. This commit softsets
persistmempool to false if zapwallettxes is enabled so transactions are
actually zapped.

This PR also fixes the zapwallettxes.py functional test, which did not properly test this feature. The test line:

```py
     assert_raises(JSONRPCException, self.nodes[0].gettransaction, [txid3])
     #there must be a expection because the unconfirmed wallettx0 must be gone by now
```
is not actually testing the presence of the transaction since the RPC is being called incorrectly (with an array instead of a string). The `assert_raises()` passes since an assert is raised, but it's not the one the test writer had in mind!

Fixes #9710 .

### CODE REVIEW

**Comment by jonasschnelli** [comment]
utACK cfc234bf063ee2c3598e3638252dd7491d61cb49.

**Comment by jnewbery** [comment]
rebased

**Comment by jnewbery** [comment]
Long-term, I think zapwallettxes should be a function in a standalone wallet tool (#8745). That's probably some time off, and in the mean time this is an improvement in behaviour. Any feedback on concept or implementation welcomed!

**Comment by MarcoFalke** [comment]
utACK 15e72f5

**Comment by sipa** [comment]
utACK 15e72f57081b2aedd3dc488364d00f0d7c99ffe0. I haven't reviewed the test changes.

**Comment by MarcoFalke** [comment]
Let's assume `persistmempool` is set in bitcoin.conf. Should there be an InitWarning when the user tries to set`zapwallettxes`?

**Comment by jnewbery** [comment]
> Let's assume persistmempool is set in bitcoin.conf. Should there be an InitWarning when the user tries to setzapwallettxes?

Perhaps, although this is an unlikely scenario since:
- persistmempool is a new in 0.15, so no-one can be using it yet
- the default value for persistmempool is true, so there's no reason to set it in bitcoin.conf

I'm inclined not to add the warning, but let me know if you feel strongly about it, and I can look at adding it.

**Comment by jnewbery** [comment]
rebased

**Comment by MarcoFalke** [comment]
Needs rebase

**Comment by jnewbery** [comment]
rebased

**Comment by achow101** [comment]
utACK 46d9a4b151cfb88011029af1310e87c5cbbd87ec

**Comment by jnewbery** [comment]
Changed `subprocess.run()` to `subprocess.check_output()` to be compatible with pre3.5 python

**Comment by morcos** [comment]
utACK 46d9a4b

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -4010,6 +4010,11 @@ bool CWallet::ParameterInteraction()
         LogPrintf("%s: parameter interaction: -salvagewallet=1 -> setting -rescan=1\n", __func__);
     }
 
+    // -zapwallettx implies dropping the mempool on startup
+    if (GetBoolArg("-zapwallettxes", false) && SoftSetBoolArg("-persistmempool", false)) {
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: "=\<mode\>" seems like a super strange string when \<mode\> isnt replaced out in the logs (plus its a boolean, what does \<mode\> mean?). zapwallettxes looks like the only one where it looks like this.

**Review by TheBlueMatt — COMMENTED** [review]
Didnt look too closely at the tests, but they look sane. Aside from the strange log message, utACK 46d9a4b151cfb88011029af1310e87c5cbbd87ec

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (jnewbery) [inline src/wallet/wallet.cpp]: Yes, you're right - that is strange. I just copied the existing log below.

I've fixed them both so they actually print the value of zapwallettxes.

Oh, and surprise! zapwallettxes isn't a bool!

**Comment by sipa** [comment]
Needs rebase.

**Comment by jnewbery** [comment]
rebased

**Comment by TheBlueMatt** [comment]
utACK 4c3b538c61532dc68d79bbe34729759a13b73f0c

**Comment by MarcoFalke** [comment]
re-utACK 4c3b538

**Pull request merged**